### DISCUSSION TITLE
Synonym Replacement와 Back Translation 결과를 공유합니다.

### DISCUSSION THREAD

**Discussion by ymnseol** [boostcampaitech4nlp2/level1_semantictextsimilarity_nlp-level1-nlp-13#43] [Show and tell]
😞 그다지 도움이 될 글은 아닌듯 싶습니다... 도움이 되지 못해 미안합니다.

아래 두 가지 방법으로 data augmentation을 시도했습니다.

1. Synonym Replacement Using BERT
2. Back Translation Using Pororo

### Overall
| Base model | Earlystopping | Loss function | Optimizer | LR scheduler | Epochs | Batch size | Initial LR |
|-------------|---------------|---------------|-----------|-----------|--------|------------|--------------------|
| klue/roberta-large | True | MAE | Adam | ReduceLROnPlateau | 10 | 64 | 1e-5 |

### Earlystopping
| Patience |
|-|
| 3 |

### Scheduler
| Factor | Patience |
|-|-|
| 0.2 | 5 |

### 1.1. 왜 SR을 수행했나요?
SR(Synonym Replacement)은 문장의 단어를 유의어로 대체하는 작업입니다. 동일한 단어가 사용되지 않았더라도 동일한 의미를 가지는 문장의 유사도를 낮게 예측하는 모델의 문제를 개선하기 위해, 학습 데이터의 문장의 일부 단어를 유의어로 바꾸어 학습시키고자 합니다.

### 1.2. 수행 과정
1. 원본 문장 + 원본 문장에서 몇몇 단어를 masking한 문장 sequence를 BERT에 넣습니다. Masking할 단어를 선택하는 방법은 한국어 형태소 분석기 'Mecab'을 사용해 형태소를 추출한 뒤, 보통명사(NNG)와 동사(VV)에 한해 특정 비율만큼 무작위로 형태소를 선택하는 방법을 사용하고 있습니다.
2. BERT가 masking된 부분에 들어갈 단어를 확률과 함께 제시합니다. (Huggingface의 `pipeline(’fill-mask’, 'klue/bert-base')` 사용) 이때 원본 문장과 다른 단어이면서, 등장 확률이 가장 높은 단어를 채택해 유의어로 대체된 문장을 한국어 문법에 맞도록 처리한 뒤 데이터에 추가합니다.

예를 들어, (데이터셋 유출 방지를 위해, 예시 문장은 데이터셋에 포함되어있지 않은 문장으로 대체했습니다.)
- 원본 문장: ‘AI가 연구와 지식에 그치지 않고 보편적인 기술이 되려면 학계와 산업 현장의 최전방에서 달리고 있는 최고 실력자들과 경험 공유가 필수입니다.’
- 마스킹된 문장: ‘AI가 연구와 지식에 그치지 않고 [MASK]적인 기술이 되려면 학계와 산업 현장의 최전방에서 달리고 있는 최고 [MASK]자들과 경험 공유가 [MASK]입니다.'
- BERT에 입력되는 문장: ‘[CLS] AI가 연구와 [MASK]에 그치지 않고 보편적인 기술이 되려면 학계와 산업 현장의 최전방에서 달리고 있는 최고 실력자들과 경험 공유가 필수입니다. [SEP] AI가 연구와 [MASK]에 그치지 않고 [MASK]적인 기술이 되려면 학계와 산업 현장의 최전방에서 달리고 있는 최고 [MASK]자들과 경험 공유가 [MASK]입니다.'
- 대체된 문장: 'AI가 연구와 학문에 그치지 않고 일상적인 기술이 되려면 학계와 산업 현장의 최전방에서 달리고 있는 최고 능력자들과 경험 공유가 관건입니다.'

원본 데이터셋과 변형된 문장쌍을 합쳤습니다. 그러므로 데이터의 크기는 기존의 2배가 됩니다.

### 1.3. 결과
⚠️  현재 리팩토링된 코드를 실행할 때 Segmentation fault (core dumped)가 발생해서, 이전 코드로 학습을 진행했습니다.
(Disc.) klue/roberta-large (sr)
<img width="1054" alt="스크린샷 2022-11-01 오후 2 13 24" src="https://user-images.githubusercontent.com/105059564/199163772-2a93d4f3-6b60-46ce-a6bf-156cee60bcdb.png">

SR로 해결되지 않은 유사도 문제를 담고 있는 문장들은 164, 172, 391 등이 있습니다. (데이터셋 유출 방지를 위해, 번호만 공개합니다. Table에서 확인 가능합니다.)
아직 val loss 변화가 불안정해보입니다. Train loss와 valid loss의 차이가 꽤 크기 때문에, '데이터의 개수가 늘어났으니까 epoch도 늘리자'라고 섣불리 결정하지 어려워 보입니다.

#### 1.4.1. SR 자체의 문제
이 방법의 가장 치명적인 문제점은 **반의어 등 적절하지 않은 단어의 등장 확률을 높게 책정하여 그 단어로 대체하는 경우가 발생**한다는 것입니다. 아무리 참고할 수 있는 원본 문장을 masking한 문장 앞에 제시한다고 해도, 반의어가 높은 등장 확률을 가지는 경우가 발생했습니다.
#### 1.4.2. 학습 과정에서의 문제
(TODO)

### 2.1. 왜 BT를 수행했나요?
앞선 BERT를 이용한 SR은 유의어가 아닌 단어가 등장할 확률이 높게 책정되어, 원본 문장과 대체 문장의 의미가 멀어질 수 있는 문제점이 있었습니다. 한국어 -> 영어 -> 한국어로의 번역을 통해, 원본 문장의 의미를 유지하는 문장을 생성하고자 합니다.

### 2.2. 수행 과정
1. 원본 문장을 영어로 번역합니다. (`mt = Pororo(task='translation', lang='multi')` -> `en_s = mt(s, src='ko', tgt='en')`)
2. 영어로 번역된 문장을 다시 한국어로 번역합니다. (`ko_s = mt(en_s, src='en', tgt='ko'`)

원본 데이터셋과 변형된 문장쌍을 합쳤습니다. 그러므로 데이터의 크기는 기존의 2배가 됩니다.

### 2.3. 결과
⚠️  현재 리팩토링된 코드를 실행할 때 Segmentation fault (core dumped)가 발생해서, 이전 코드로 학습을 진행했습니다.

(Disc.) klue/roberta-large (bt)

<img width="916" alt="스크린샷 2022-11-01 오후 3 03 32" src="https://user-images.githubusercontent.com/105059564/199169284-68ee4851-5ea1-4f25-8b82-71cf6cf78338.png">

Train loss가 기존 최고 성능 모델 및 1번에서의 모델과 비교했을 때 느리게 수렴하는 것을 볼 수 있습니다. 10 epochs 돌렸는데 아직 val loss가 비교적 안정적으로 감소하고 있어서, 더 실험을 해보아야 할 것 같습니다.

#### 2.4.1. BT 자체의 문제
Pororo의 BT 결과가 생각보다 자연스럽지 않습니다. 그래서 BT를 한 뒤 원본 문장과 번역 문장의 유사도를 비교하여 높은 문장들만 데이터셋에 추가하는 방식도 사용한다고 합니다. 우선은 이 방법까지는 적용하지 않았습니다. **'현재 BERT 기반 우리 모델이 단어 의존 문제 때문에 문장간 유사도를 제대로 측정하지 못하는데, 과연 원본 문장과 번역 문장을 비교했을 때 유사도를 잘 측정할까'라는 의구심**이 들었기 때문입니다. 이 부분은 시간이 있다면 실험을 통해 알아보고자 합니다.

### 3.1. 기존 데이터 기반 pre-trained model로 fill-mask
준휘오빠가 공유한 MLM 모델로 fill-mask해보기
### 3.2. Cross validation
Train loss와 val loss의 차이를 보았을 때, generalization performance가 떨어진다는 것을 알 수 있습니다. 성구오빠 말대로, cross validation을 적용해보는 것도 좋을듯합니다. 다만, sequence data의 cross validation을 어떻게 해야할지, 지난 강의에서 언급한 방법들을 검토해보고 구현하는 데에 시간이 걸리지 않을까 싶습니다.
### 3.3. KoSimCSE
우선은 현재 모델을 바꾸지 않는 선에서 개선을 하는 게 가장 좋지 않을까 싶지만...  
그래도 단어 의존 문제에서 비교적 자유로운 모델을 사용해보고자 한다면, KoSimCSE 사용을 시도해볼 수 있겠습니다. 만약 KoSimCSE가 기존에 사용하던 모델만큼의 성능을 내지 못한다면, KoSimCSE가 단어 의존 문제를 해소하는지 살펴보고, 해소한다면 기존 모델과 앙상블해서 사용하는 방법을 시도해볼 수 있지 않을까 싶습니다.

**Comment by intrandom5** [comment]
수고했다... SimCSE 같은 경우에는 내가 지금 논문 읽어 봤고, 이제 우리 데이터셋에 어떻게 적용해볼까 생각하는 참이었어.
우리 노션에 정리글 올렸으니까 이걸로 한 번 봐봐
https://www.notion.so/SimCSE-79699e63d59c431fa1158442b062c974
